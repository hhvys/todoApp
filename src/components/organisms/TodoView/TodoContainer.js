import React from 'react';
import TodoView from "./TodoView";
import {SORT_BY} from "../../../actions/actionTypes";

class TodoContainer extends React.Component {

	constructor(props) {
		super(props);
		this.state = {
			currentTime: new Date()
		};
	}

	componentDidMount() {
		this.interval = setInterval(() => {
			this.setState({
				currentTime: new Date()
			});
		}, 65000);
	}

	componentWillUnmount() {
		clearInterval(this.interval);
	}

	sortTodos = (todos, sortBy) => (
		todos
			.sort((a, b) => {
				switch (sortBy) {
					case SORT_BY.SORT_PRIORITY:
						if (a.star && !b.star)
							return -1;
						if (!a.star && b.star)
							return 1;
						return 0;

					case SORT_BY.SORT_CREATION:
						if (a.completedTime > b.completedTime)
							return -1;
						return 1;

					case SORT_BY.SORT_ALPHA:
						if (a.text < b.text)
							return -1;
						return 1;
				}

			})
	);

	getCompletionTime = (completedTime) => {
		let timeDiff = Math.abs(this.state.currentTime - Date.parse(completedTime)) / 1000;

		switch (true) {
			case (timeDiff < 60):
				return 'A few seconds ago';
			case (timeDiff < 120):
				return 'A minute ago';
			case (timeDiff < 3600):
				return Math.ceil((timeDiff - 60) / 60).toString() + ' minutes ago';
			case (timeDiff < 7200):
				return 'An hour ago';
			default:
				return Math.ceil((timeDiff - 3600) / 3600).toString() + ' hours ago';
		}
	};

	addCompletionTime = (todos) => (
		todos
			.map(todo => ({
				...todo,
				completionTime: this.getCompletionTime(todo.completedTime)
			}))
	);

	render() {
		let {todos, sortBy, ...props} = this.props;
		todos = this.sortTodos(todos, sortBy);
		todos = this.addCompletionTime(todos);
		return (
			<TodoView
				todos={todos}
				{...props}
			/>
		);
	}
}

export default TodoContainer;
