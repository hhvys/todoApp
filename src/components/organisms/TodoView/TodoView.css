.div__todo__view {
    margin-left: 300px;
    margin-top: 70px;
    height: 2px;
    height: 100%;
}

