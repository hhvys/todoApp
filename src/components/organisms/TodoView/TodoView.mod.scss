@import "../../../constants.mod";

.todoView {
  margin-left: ($side-bar-width + 20px);
  margin-top: ($header-height+25)+px;
  height: 100%;
}

.collapsed {
  margin-left: ($collpased-side-bar-width + 20px);
}

.headerLogo {
  margin-left: 10px;
}

.footerLogo {
  margin-right: 10px;
  fill: rgba(0, 0, 0, 0.3);
}

.todo {
  margin-top: 2px;
  background-color: #F7F7F7;
  border-radius: 5px;
  height: $todo-height+px;
}

.completedTodo {
  @extend .todo;
  opacity: 0.6;
}

.input {
  font-size: 16px;
}
