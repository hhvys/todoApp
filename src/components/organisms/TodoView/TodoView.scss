@import "../../../constants";

.todoView {
  margin-left: ($side-bar-width + 20px);
  margin-top: 70px;
  height: 100%;
}

.collapsed {
  margin-left: 62px;
}

.headerLogo {
  margin-left: 10px;
}

.footerLogo {
  margin-right: 10px;
  fill: rgba(0, 0, 0, 0.3);
}

.todo {
  margin-top: 2px;
  background-color: #F7F7F7;
  border-radius: 5px;
  height: 46px;
}

.completedTodo {
  @extend .todo;
  opacity: 0.6;
}

.input {
  font-size: 16px;
}
