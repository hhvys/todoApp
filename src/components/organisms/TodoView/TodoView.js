import React from 'react';
import './TodoView.css';
import InputWithLabel from "../../molecules/InputWithLabel/InputWithLabel";
import VerticalTab from "../../molecules/VerticalTab/VerticalTab";
import Button from "../../atoms/Button/Button";
import {CHECK_BOX, CHECKED_CHECK_BOX, STARRED, STAR} from "../../atoms/logos/constants";
import {SORT_BY} from "../../../actions/actionTypes";

class TodoView extends React.Component {

	constructor(props) {
		super(props);
		this.state = {
			currentTime: new Date()
		};
	}

	componentDidMount() {
		this.interval = setInterval(() => {
			this.setState({
				currentTime: new Date()
			});
		}, 65000);
	}

	componentWillUnmount() {
		clearInterval(this.interval);
	}


	getCompletionTime = (completedTime) => {
		let timeDiff = Math.abs(this.state.currentTime - Date.parse(completedTime)) / 1000;

		switch (true) {
			case (timeDiff < 60):
				return 'A few seconds ago';
			case (timeDiff < 120):
				return 'A minute ago';
			case (timeDiff < 3600):
				return Math.ceil((timeDiff - 60) / 60).toString() + ' minutes ago';
			case (timeDiff < 7200):
				return 'An hour ago';
			default:
				return Math.ceil((timeDiff - 3600) / 3600).toString() + ' hours ago';
		}
	};

	renderVerticalTab = ({
												 todos,
												 activeTab,
												 onFooterSymbolClick,
												 onHeaderSymbolClick,
												 completed,
												 sortBy
											 }) => {

		return (
			todos
				.sort((a, b) => {
					switch (sortBy) {
						case SORT_BY.SORT_PRIORITY:
							if (a.star && !b.star)
								return -1;
							if (!a.star && b.star)
								return 1;
							return 0;

						case SORT_BY.SORT_CREATION:
							if (a.completedTime > b.completedTime)
								return -1;
							return 1;

						case SORT_BY.SORT_ALPHA:
							if (a.text < b.text)
								return -1;
							return 1;
					}

				})
				.map(todo => {

					return (
						<VerticalTab key={todo.todoId}
												 headerSymbol={todo.completed ? CHECKED_CHECK_BOX : CHECK_BOX}
												 footerSymbol={todo.star ? STARRED : STAR}
												 headerProps={{
													 style: {
														 marginLeft: 10
													 }
												 }}
												 footerProps={{
													 style: {
														 marginRight: 10,
												 }
												 }}
												 mainContent={todo.text}
												 style={{
													 marginTop: 2,
													 backgroundColor: '#F7F7F7',
													 borderRadius: '5px',
													 height: 46,
													 opacity: completed ? 0.6 : 1
												 }}

												 onFooterSymbolClick={() => onFooterSymbolClick(activeTab, todo.todoId)}
												 onHeaderSymbolClick={() => onHeaderSymbolClick(activeTab, todo.todoId)}
												 extraContent={completed ? this.getCompletionTime(todo.completedTime) : null}
						/>
					);
				})
		);
	};

	render() {
		const {
			collapsed,
			className,
			todos,
			activeTab,
			onFooterSymbolClick,
			onHeaderSymbolClick,
			onButtonClick,
			showCompleted,
			onInputSubmit,
			sortBy,
			...props
		} = this.props;

		return (
			<div
				className={`${className ? className : ''} ${collapsed ? 'collapsed' : ''} div__todo__view full-height d-flex flex-column justify-content-start mr-3`} {...props}>
				<InputWithLabel
					label={"+"}
					placeholder={"Add a to-do..."}
					style={{fontSize: 16}}
					onSubmit={(value) => onInputSubmit(activeTab, value)}
				/>

				<div className={"incomplete-todos mt-3"}>
					{this.renderVerticalTab({
						todos: todos.filter(todo => !todo.completed),
						activeTab,
						onFooterSymbolClick,
						onHeaderSymbolClick,
						completed: false,
						sortBy
					})}

				</div>

				<Button className={"mt-3"}
								text={showCompleted ? "HIDE COMPLETED TO-DOS" : "SHOW COMPLETED TO-DOS"}
								onClick={() => onButtonClick(activeTab)}/>

				<div className={"completed-todos mt-3"}>
					{showCompleted && this.renderVerticalTab({
						todos: todos.filter(todo => todo.completed),
						activeTab,
						onFooterSymbolClick,
						onHeaderSymbolClick,
						completed: true,
						sortBy,
					})}
				</div>


			</div>

		);
	}
}

export default TodoView;
