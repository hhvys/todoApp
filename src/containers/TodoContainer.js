import {connect} from 'react-redux';
import TodoView from '../components/organisms/TodoView/TodoView';
import {addTodo, toggleStarTodo, toggleTodo, toggleVisibilityFilter} from "../actions/actionCreaters";
import {getTabs} from "../reducers/tabs";
import {getActiveTab} from "../reducers/activeTab";
import {getSortBy} from "../reducers/sortBy";
import {getCollapsedSideBar} from "../reducers/collapsedSideBar";

const mapStateToProps = (state) => {
	const tabs = getTabs(state);
	const activeTab = getActiveTab(state);
	const todos = tabs
		.find(tab => tab.tabId === activeTab)
		.todos;
	return ({
		todos,
		activeTab,
		showCompleted: tabs
			.find(tab => tab.tabId === activeTab)
			.showCompletedTodo,
		sortBy: getSortBy(state),
		collapsed: getCollapsedSideBar(state)
	});
};

const mapDispatchToProps = (dispatch) => {
	return ({
		onHeaderSymbolClick: (tabId, todoId) => dispatch(toggleTodo(tabId, todoId)),
		onFooterSymbolClick: (tabId, todoId) => dispatch(toggleStarTodo(tabId, todoId)),
		onButtonClick: (tabId) => dispatch(toggleVisibilityFilter(tabId)),
		onInputSubmit: (tabId, value) => dispatch(addTodo(tabId, value))
	});
};

const TodoContainer = connect(
	mapStateToProps,
	mapDispatchToProps
)(TodoView);

export default TodoContainer;
