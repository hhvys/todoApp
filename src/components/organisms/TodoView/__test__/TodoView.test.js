import React from 'react';
import TodoView from '../TodoView';
import TodoContainer from "../TodoContainer";

describe('components', () => {
	describe('TodoView', () => {
		let props;
		let mountedTodoView;

		const todoView = () => {
			if (!mountedTodoView) {
				mountedTodoView = shallow(
					<TodoView {...props}/>
				);
			}
			return mountedTodoView;
		};

		beforeEach(() => {
			props = {
				collapsed: false,
				className: "className",
				todos: [],
				activeTab: undefined,
				onFooterSymbolClick: jest.fn(),
				onHeaderSymbolClick: jest.fn(),
				onButtonClick: jest.fn(),
				showCompleted: false,
				onInputSubmit: jest.fn(),
			};
			mountedTodoView = undefined;
		});

		it('should have collapsed class if collapsed is true', () => {
			props.collapsed = true;
			const wrapperDiv = todoView().find('div').first();
			expect(wrapperDiv.props().className).toMatch('collapsed');
		});

		it('should not have collapsed class if collapsed is false', () => {
			props.collapsed = false;
			const wrapperDiv = todoView().find('div').first();
			expect(wrapperDiv.props().className).not.toMatch('collapsed');
		});

		it('should have number of RowComponents = todos.length if showCompleted is true', () => {
			const RowComponents = todoView().find('RowComponent');
			expect(RowComponents.length).toBe(props.todos.length);
		});

		it('should have number of RowComponents = uncompleted todos.length if showCompleted is false', () => {
			const wrapperDiv = todoView().find('div').first();
			const completedTodoDiv = wrapperDiv.find('div').last();
			const RowComponents = completedTodoDiv.find('RowComponent');
			const uncompletedTodos = props.todos.filter(todo => todo.completed).length;
			expect(RowComponents.length).toBe(uncompletedTodos);
		});

		describe('InputWithLabel', () => {
			it('should render InputWithLabel', () => {
				const InputWithLabel = todoView().find('InputWithLabel');
				expect(InputWithLabel.length).toBe(1);
			});
		});

		describe('Button', () => {
			it('should render Button', () => {
				expect(todoView()).toMatchSnapshot();
			});
		});

	});

	describe('TodoContainer', () => {

		describe('getCompletionTime', () => {

			it('should return "A few seconds ago" ', () => {
				const completedTime = new Date();
				const todoContainer = new TodoContainer({});
				todoContainer.state = {
					currentTime: new Date(completedTime.getTime() + 15 * 1000)
				};
				expect(todoContainer.getCompletionTime(completedTime)).toBe('A few seconds ago');
			});

			it('should return "A minute ago" ', () => {
				const completedTime = new Date();
				const todoContainer = new TodoContainer({});
				todoContainer.state = {
					currentTime: new Date(completedTime.getTime() + 70 * 1000)
				};
				expect(todoContainer.getCompletionTime(completedTime)).toBe('A minute ago');
			});

			it('should return "15 minutes ago" ', () => {
				const completedTime = new Date();
				const todoContainer = new TodoContainer({});
				todoContainer.state = {
					currentTime: new Date(completedTime.getTime() + 15 * 60000)
				};
				expect(todoContainer.getCompletionTime(completedTime)).toBe('15 minutes ago');
			});

			it('should return "An hour ago" ', () => {
				const completedTime = new Date();
				const todoContainer = new TodoContainer({});
				todoContainer.state = {
					currentTime: new Date(completedTime.getTime() + 65 * 60000)
				};
				expect(todoContainer.getCompletionTime(completedTime)).toBe('An hour ago');
			});

			it('should return "2 hours ago" ', () => {
				const completedTime = new Date();
				const todoContainer = new TodoContainer({});
				todoContainer.state = {
					currentTime: new Date(completedTime.getTime() + 145 * 60000)
				};
				expect(todoContainer.getCompletionTime(completedTime)).toBe('2 hours ago');
			});

		});

	});

});

