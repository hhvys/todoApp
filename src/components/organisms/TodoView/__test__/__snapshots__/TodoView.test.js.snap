// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components TodoView Button should render Button 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoView
    activeTab={undefined}
    className="className"
    collapsed={false}
    onButtonClick={[Function]}
    onFooterSymbolClick={[Function]}
    onHeaderSymbolClick={[Function]}
    onInputSubmit={[Function]}
    showCompleted={false}
    todos={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <InputWithLabel
          className="input"
          label="+"
          onSubmit={[Function]}
          placeholder="Add a to-do..."
/>,
        <div
          className="mt-3"
>
          
</div>,
        <Button
          className="mt-3"
          onClick={[Function]}
          text="SHOW COMPLETED TO-DOS"
/>,
        <div
          className="mt-3"
/>,
      ],
      "className": "className  todoView full-height d-flex flex-column justify-content-start mr-3",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "input",
          "label": "+",
          "onSubmit": [Function],
          "placeholder": "Add a to-do...",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "mt-3",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "mt-3",
          "onClick": [Function],
          "text": "SHOW COMPLETED TO-DOS",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": false,
          "className": "mt-3",
        },
        "ref": null,
        "rendered": false,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <InputWithLabel
            className="input"
            label="+"
            onSubmit={[Function]}
            placeholder="Add a to-do..."
/>,
          <div
            className="mt-3"
>
            
</div>,
          <Button
            className="mt-3"
            onClick={[Function]}
            text="SHOW COMPLETED TO-DOS"
/>,
          <div
            className="mt-3"
/>,
        ],
        "className": "className  todoView full-height d-flex flex-column justify-content-start mr-3",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "input",
            "label": "+",
            "onSubmit": [Function],
            "placeholder": "Add a to-do...",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "mt-3",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "mt-3",
            "onClick": [Function],
            "text": "SHOW COMPLETED TO-DOS",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": false,
            "className": "mt-3",
          },
          "ref": null,
          "rendered": false,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
